{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Genomes rank analysis and get genomes RefSeq accession IDs \n",
    "\n",
    "#Load modules\n",
    "import pandas as pd\n",
    "import argparse\n",
    "import os\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download NCBI taxonomy IDs and ranks \n",
    "def download_taxonomy():\n",
    "    names_NCBI = {} #names as keys and NCBI IDs as values\n",
    "    nodes_NCBI = {} #NCBI IDs as keys and ranks as values\n",
    "    ranks = set()\n",
    "\n",
    "    with open('./names.dmp', 'r') as file: \n",
    "        for line in file:\n",
    "            line = line.strip().split('|')\n",
    "            line = [i.rstrip().lstrip() for i in line]\n",
    "            name = line[1]\n",
    "            id = line[0]\n",
    "            info = line[3]\n",
    "\n",
    "            if info == 'scientific name':\n",
    "                try:\n",
    "                    names_NCBI[name].append(id)\n",
    "                except:\n",
    "                    names_NCBI[name] = [id]\n",
    "\n",
    "    with open('./nodes.dmp', 'r') as file:\n",
    "        for line in file:\n",
    "            line = line.strip().split('|')\n",
    "            line = [i.rstrip().lstrip() for i in line[0:3]] \n",
    "            nodes_NCBI[line[0]] =  line[1], line[2]\n",
    "            ranks.add(line[2])\n",
    "            \n",
    "    return names_NCBI, nodes_NCBI, ranks\n",
    "\n",
    "\n",
    "#Get genome Refseq accession IDs per taxon name\n",
    "def get_genomes_by_name(genomes):\n",
    "    gen_by_name = {}\n",
    "\n",
    "    for genome in genomes:\n",
    "        list_acc_ids = []\n",
    "        for i in genomes[genome]:\n",
    "            acc_id = i[0]\n",
    "            list_acc_ids.append(acc_id)\n",
    "        gen_by_name[genome] = list_acc_ids\n",
    "\n",
    "    return gen_by_name\n",
    "\n",
    "#Get genome Refseq accession IDs per taxon ID using NCBI taxonomy\n",
    "def get_gen_ids(genomes):\n",
    "    #Call get_genomes_by_name and substitute name with NCBI ID\n",
    "    gen_by_name = get_genomes_by_name(genomes)\n",
    "    gen_by_id = {}\n",
    "\n",
    "    for name in gen_by_name:\n",
    "        genomes = gen_by_name[name]\n",
    "        for id in taxonomy[name]:\n",
    "            gen_by_id[id] = genomes\n",
    "\n",
    "    return gen_by_id\n",
    "\n",
    "#Parse file obtained from dry run of ncbi-genome-download\n",
    "def get_genomes(path):\n",
    "    genomes = {}\n",
    "    count = 0\n",
    "\n",
    "    with open(path, 'r') as file:\n",
    "        for line in file:\n",
    "            line = line.strip().split('\\t')\n",
    "            if not line[0].startswith('Considering'):\n",
    "                count += 1\n",
    "                name = line[1]\n",
    "                try: \n",
    "                    genomes[name].append([line[0], line[2]])\n",
    "\n",
    "                except KeyError:\n",
    "                    genomes[name] = [[line[0], line[2]]]\n",
    "\n",
    "    print('{} genomes present in {}'.format(count, path))\n",
    "\n",
    "    #Return list of Refseq accession ID per taxa ID\n",
    "    #dictionary with taxaID as keys and genomes as values\n",
    "    gen_by_id = get_gen_ids(genomes)\n",
    "\n",
    "    return gen_by_id\n",
    "\n",
    "\n",
    "#Retrieve IDs with right rank among the present genomes IDs or their parent IDs\n",
    "def right_rank(id, rank):\n",
    "    rank_index = my_lineage.index(rank)\n",
    "    ancestors = my_lineage[:rank_index]\n",
    "    \n",
    "    if id == '447024':\n",
    "        res = 'species'\n",
    "    else:\n",
    "        res = nodes_NCBI[id][1] #rank\n",
    "    if res in ancestors: \n",
    "        return\n",
    "    elif res == rank:\n",
    "        return id\n",
    "    else:\n",
    "        if id == '447024': \n",
    "            parent = '2614943'\n",
    "        else:\n",
    "            parent = nodes_NCBI[id][0] #parent id\n",
    "\n",
    "        return right_rank(parent,rank)\n",
    "    \n",
    "#For each ID at right rank, store list of genomes belonging to that taxa \n",
    "def ranking_analysis(rank, test):\n",
    "    data = {}\n",
    "\n",
    "    for id in test:\n",
    "        result = right_rank(id, rank)\n",
    "        if result is not None:\n",
    "            for c in range(len(test[id])):\n",
    "                try:\n",
    "                    data[result].append(id)\n",
    "                except:\n",
    "                    data[result] = [id]\n",
    "        else: continue\n",
    "\n",
    "    return data\n",
    "\n",
    "#Perform distribution analysis of genomes at given rank\n",
    "def genome_analysis(rank, genomes_id):\n",
    "\n",
    "     print('Analyse {} distribution...'.format(rank))\n",
    "     result = ranking_analysis(rank, genomes_id)\n",
    "\n",
    "     print('Create outputs...')\n",
    "     df = {}\n",
    "     df['ID'] = [id for id in result]\n",
    "     names = {i:name for name in taxonomy for i in taxonomy[name]}\n",
    "     df['Names'] = [names[id] for id in result]\n",
    "     df['Number'] = [len(result[id]) for id in result] \n",
    "\n",
    "     return pd.DataFrame(data = df).sort_values(by = 'Number', ascending=False), result\n",
    "\n",
    "\n",
    "\n",
    "def retrieve_descen_ids(descendants):\n",
    "    genomes_complete = []\n",
    "    genomes = []\n",
    "    for id in descendants:\n",
    "        g = genomes_id[id]\n",
    "        genomes += g\n",
    "\n",
    "        if id in complete_genomes_id:\n",
    "            genomes_complete += complete_genomes_id[id]\n",
    "    \n",
    "    return genomes, genomes_complete\n",
    "\n",
    "#Retrieve max n accession IDs per taxa \n",
    "def get_acc_ids(species_df, species_dict, n = 30):\n",
    "    species = species_df['ID']\n",
    "    info_genomes = {}\n",
    "    genomes_of_species = {}\n",
    "    acc_ids = []\n",
    "\n",
    "    under = list(species_df.loc[species_df['Number'] <= n]['ID']) #7868\n",
    "    up = list(species_df.loc[species_df['Number'] > n]['ID']) #84\n",
    "\n",
    "    for s in species:\n",
    "        des_set = set(species_dict[s]) #descendants\n",
    "\n",
    "        s_acc_ids, complete_acc_ids = retrieve_descen_ids(des_set)\n",
    "\n",
    "        #if taxa genomes < n then all genomes are taken\n",
    "        if s in under:\n",
    "            genomes_of_species[s] = s_acc_ids\n",
    "            info_genomes[s] = [str(len(genomes_of_species[s])) , str(len(complete_acc_ids))]\n",
    "            acc_ids += s_acc_ids\n",
    "        \n",
    "        #if taxa has too many genomes, we check for number of complete genomes\n",
    "        #if complete genomes are too many, we pick randomly \n",
    "        elif s in up and len(complete_acc_ids) > n:\n",
    "            taken_acc_ids = random.sample(complete_acc_ids, n)\n",
    "            genomes_of_species[s] = taken_acc_ids\n",
    "            info_genomes[s] = [str(len(taken_acc_ids)), str(len(complete_acc_ids))]\n",
    "            acc_ids += taken_acc_ids\n",
    "        \n",
    "        #if no complete genomes are available, we pick randomly among the not completed genomes\n",
    "        elif s in up and len(complete_acc_ids) == 0:\n",
    "            taken_acc_ids = random.sample(s_acc_ids, n)\n",
    "            genomes_of_species[s] = taken_acc_ids\n",
    "            info_genomes[s] = [str(len(taken_acc_ids)), str(len(complete_acc_ids))]\n",
    "            acc_ids += taken_acc_ids\n",
    "\n",
    "        else: #pick available complete genomes\n",
    "            genomes_of_species[s] = complete_acc_ids\n",
    "            info_genomes[s] = [str(len(complete_acc_ids)), str(len(complete_acc_ids))]\n",
    "            acc_ids += complete_acc_ids\n",
    "\n",
    "    print('{} RefSeq IDs stored...'.format(len(acc_ids)))\n",
    "\n",
    "    return acc_ids, genomes_of_species, info_genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] -i GENOME_LIST -c COMPLETE_LIST\n",
      "                             [-r {superkingdom,kingdom,subkingdom,superphylum,phylum,subphylum,infraphylum,superclass,class,subclass,infraclass,cohort,subcohort,superorder,order,suborder,infraorder,parvorder,superfamily,family,subfamily,tribe,subtribe,genus,subgenus,section,subsection,series,subseries,species group,species subgroup,species,forma specialis,subspecies,morph,varietas,subvariety,forma,serogroup,serotype,biotype,strain,pathogroup,isolate,genotype,no rank,clade}]\n",
      "                             -o OUTPUT_FOLDER [-n NUMBER_GENOMES]\n",
      "ipykernel_launcher.py: error: the following arguments are required: -i/--genome_list, -c/--complete_list, -o/--output_folder\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/baccus/.local/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3386: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "if __name__=='__main__':\n",
    "    parser = argparse.ArgumentParser(description='Genomes distribution analysis')\n",
    "    parser.add_argument('-i', '--genome_list', required=True, help=\"Filepath to list of genomes obtained from dry run (-n command) of ncbi-genome-download\")\n",
    "    parser.add_argument('-c', '--complete_list', required=True, help=\"Filepath to list of complete genomes obtained from dry run (-n command) of ncbi-genome-download\")\n",
    "    parser.add_argument('-r', '--rank', default= 'species', choices =['superkingdom', 'kingdom', 'subkingdom', 'superphylum', 'phylum', 'subphylum', 'infraphylum', 'superclass', 'class', 'subclass', 'infraclass', 'cohort', 'subcohort', 'superorder', 'order', 'suborder', 'infraorder', 'parvorder', 'superfamily', 'family', 'subfamily', 'tribe', 'subtribe', 'genus', 'subgenus', 'section', 'subsection', 'series', 'subseries', 'species group', 'species subgroup','species', 'forma specialis', 'subspecies', 'morph','varietas', 'subvariety', 'forma', 'serogroup', 'serotype', 'biotype','strain', 'pathogroup','isolate', 'genotype', 'no rank', 'clade'],help = \"Taxonomic rank whose distribution among genomes must be analysed, species by default.\" )\n",
    "    parser.add_argument('-o', '--output_folder', required = True, help = \"Output folder. Outputs include .csv files to analyse the distribution of desired rank in given genomes, accession_ID.txt with list of RefSeq genome accession IDs useful to download genomes from NCBI. \")\n",
    "    parser.add_argument('-n', '--number_genomes', type=int, help = \"Maximum number of genomes to be downloaded per taxon.\")\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print('Download taxonomy from NCBI...')\n",
    "    taxonomy, nodes_NCBI, ranks = download_taxonomy()\n",
    "        \n",
    "    #if output folder does not exist, create it\n",
    "    if os.path.exists(args.output_folder) == False:\n",
    "        os.makedirs(args.output_folder)\n",
    "\n",
    "    print('Retrieving genomic information from provided data...')\n",
    "    genomes_id = get_genomes(args.genome_list)\n",
    "    complete_genomes_id = get_genomes(args.complete_list)\n",
    "\n",
    "    #NCBI taxonomic rank hierarchy\n",
    "    my_lineage = ['superkingdom', 'kingdom', 'subkingdom', 'superphylum', 'phylum', 'subphylum', 'infraphylum', 'superclass', 'class', 'subclass', 'infraclass', 'cohort', 'subcohort', 'superorder', 'order', 'suborder', 'infraorder', 'parvoorder', 'superfamily', 'family', 'subfamily', 'tribe', 'subtribe', 'genus', 'subgenus', 'section', 'subsection', 'series', 'subseries', 'species group', 'species subgroup','species', 'forma specialis', 'subspecies', 'morph','varietas', 'subvariety', 'forma', 'serogroup', 'serotype', 'biotype','strain', 'pathogroup','isolate', 'genotype', 'no rank', 'clade']\n",
    "\n",
    "    results_df, result_dict = genome_analysis(args.rank, genomes_id)\n",
    "    \n",
    "    print('{} {} detected'.format(len(result_dict), args.rank))\n",
    "\n",
    "    results_df.to_csv(os.path.join(args.output_folder,'id_name_number.csv'))\n",
    "    counts_df = pd.DataFrame(data={'Species': results_df.groupby(['Number']).count()['Names']})\n",
    "    counts_df['Percentage'] = counts_df['Species']/counts_df['Species'].sum()*100\n",
    "    counts_df.to_csv(os.path.join(args.output_folder,'genome_numbers.csv'))\n",
    "\n",
    "    if args.number_genomes:\n",
    "        print('Retrieve {} RefSeq accession IDs per {}... '.format(args.number_genomes, args.rank))\n",
    "        acc_ids, genomes_of_species, info_genomes = get_acc_ids(results_df,result_dict, args.number_genomes)\n",
    "\n",
    "\n",
    "        with open(os.path.join(args.output_folder,'accession_ID.txt'), 'w') as file:\n",
    "            file.writelines([id + '\\n' for id in acc_ids])\n",
    "            file.close()\n",
    "\n",
    "        with open(os.path.join(args.output_folder,'genomes_per_species.txt'), 'w') as file:\n",
    "            file.write('Species : NCBI accession IDs per species \\n')\n",
    "            out = [s + '\\t:\\t' + ' '.join(genomes_of_species[s]) + '\\n' for s in genomes_of_species]\n",
    "            file.writelines(out)\n",
    "            file.close()\n",
    "\n",
    "        with open(os.path.join(args.output_folder,'genomes_info.txt'), 'w') as file:\n",
    "            file.write('Species : Number of genomes, Number of complete genomes \\n')\n",
    "            file.writelines([str(s) + '\\t:\\t' + ' '.join(info_genomes[s]) + '\\n' for s in info_genomes])\n",
    "            file.close()\n",
    "\n",
    "        with open(os.path.join(args.output_folder,'genomes_id.txt'), 'w') as file:\n",
    "            file.write('NCBI taxaID : NCBI accession IDs \\n')\n",
    "            out = [s + '\\t:\\t' + ' '.join(genomes_of_species[s]) + '\\n' for s in genomes_of_species]\n",
    "            file.writelines(out)\n",
    "            file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
